### PR TITLE
bai 5 menu date và time picker

### PR DESCRIPTION
**Pull request by Ngduong29**
## Related Tickets
- [#Task xxxxx: Content](https://edu-redmine.sun-asterisk.vn/issues/xxxx)
## WHAT
- Bài tập slide menu và picker
## Evidence (Screenshot or Video)
image
image
image
image
image
image
image
image
image
image

## Review Checklist

Category | View Point | Description | Expected Reviewer Answer | Self review | Reviewer2 (name)
--- | --- | --- | --- | --- | ---
Conventions | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/blob/master/eng/android/coding_convention_android_kotlin.md | YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>
Redmine | Does the ticket follow Sun* Redmine working process?  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md| YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>
Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |  | YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>
## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(Other notes)*

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -10,33 +10,52 @@
...
         </activity>
         <activity
             android:name=".MainActivity"
-            android:exported="false">
-
-        </activity>
+            android:exported="false" />
     </application>
 
 </manifest>
```
Comment (hungvv-1301) [inline app/src/main/AndroidManifest.xml]:
```suggestion
</manifest>

```

File: app/src/main/java/com/sun/android/menus_picker/AlertActivity.kt [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+
+class AlertActivity : AppCompatActivity() {
+    val binding by lazy { ActivityAlertBinding.inflate(layoutInflater) }
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(binding.root)
+        binding.button.setOnClickListener{
+            var myAlertBuilder: AlertDialog.Builder = AlertDialog.Builder(this)
+            myAlertBuilder.setTitle("Alert")
```
Comment (hungvv-1301) [inline app/src/main/java/com/sun/android/menus_picker/AlertActivity.kt:18]: extract string

File: app/src/main/java/com/sun/android/menus_picker/AlertActivity.kt [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+            myAlertBuilder.setMessage("Click OK to continue, or Cancel to stop: ")
+            myAlertBuilder.setPositiveButton("OK") { dialog, which ->
+                Toast.makeText(
+                    applicationContext, "Pressed OK",
+                    Toast.LENGTH_SHORT
+                ).show()
+            }
+            myAlertBuilder.setNegativeButton("Cancel") { dialog, which ->
+                Toast.makeText(
+                    applicationContext, "Pressed Cancel",
```
Comment (hungvv-1301) [inline app/src/main/java/com/sun/android/menus_picker/AlertActivity.kt]: extract string

File: app/src/main/java/com/sun/android/menus_picker/MainActivityMenuPicker.kt [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+import android.view.Menu
+import android.view.MenuItem
+import android.view.View
+import android.widget.Toast
+import androidx.appcompat.app.AppCompatActivity
+import com.sun.android.EXTRA_MESSEAGE
+import com.sun.android.R
+import com.sun.android.databinding.ActivityMainMenuPickerBinding
+
+var mOrderMessage: String = ""
```
Comment (hungvv-1301) [inline app/src/main/java/com/sun/android/menus_picker/MainActivityMenuPicker.kt]: để vào trong class

File: app/src/main/java/com/sun/android/menus_picker/PickDateActivity.kt [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        }
+
+
+    }
+    fun processDatePickerResult(year: Int, month: Int, day: Int){
+        val month_string = Integer.toString(month + 1)
+        val day_string = Integer.toString(day)
+        val year_string = Integer.toString(year)
+        val dateMessage = day_string + "/" + month_string + "/" + year_string
+        Toast.makeText(this, "Date: " + dateMessage, Toast.LENGTH_SHORT).show()
```
Comment (hungvv-1301) [inline app/src/main/java/com/sun/android/menus_picker/PickDateActivity.kt]: extract string

Comment (hungvv-1301) [reply app/src/main/java/com/sun/android/menus_picker/PickDateActivity.kt]: dùng %s %d để fill

File: app/src/main/res/layout/activity_alert.xml [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginBottom="@dimen/dp_26"
+        android:text="@string/alert"
+        app:layout_constraintBottom_toBottomOf="parent"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toBottomOf="@+id/textView"
+        />
+</androidx.constraintlayout.widget.ConstraintLayout>
```
Comment (hungvv-1301) [inline app/src/main/res/layout/activity_alert.xml]:
```suggestion
</androidx.constraintlayout.widget.ConstraintLayout>

```

File: app/src/main/res/layout/activity_alert.xml [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+        android:text="@string/alertText"
+        app:layout_constraintBottom_toTopOf="@+id/button"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent"
+        app:layout_constraintVertical_chainStyle="packed"
+        />
+
+    <Button
+        android:id="@+id/button"
```
Comment (hungvv-1301) [inline app/src/main/res/layout/activity_alert.xml]: button nên có tên

File: app/src/main/res/layout/activity_main_menu_picker.xml [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+        </androidx.constraintlayout.widget.ConstraintLayout>
+
+    </androidx.core.widget.NestedScrollView>
+
+    <com.google.android.material.floatingactionbutton.FloatingActionButton
+        android:id="@+id/fab"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_gravity="bottom|end"
+        android:layout_margin="16dp"
```
Comment (hungvv-1301) [inline app/src/main/res/layout/activity_main_menu_picker.xml]: dimen

File: app/src/main/res/layout/activity_main_menu_picker.xml [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+        app:srcCompat="@drawable/ic_shopping_cart" />"
+
+</androidx.coordinatorlayout.widget.CoordinatorLayout>
```
Comment (hungvv-1301) [inline app/src/main/res/layout/activity_main_menu_picker.xml]:
```suggestion
</androidx.coordinatorlayout.widget.CoordinatorLayout>

```

File: app/src/main/res/layout/activity_order.xml [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+
+    <TextView
+        android:id="@+id/order_textview"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginLeft="24dp"
+        android:layout_marginStart="24dp"
+        android:layout_marginTop="8dp"
+        android:text="@string/order_label_text"
+        android:textSize="18sp"
```
Comment (hungvv-1301) [inline app/src/main/res/layout/activity_order.xml]: .

File: app/src/main/res/layout/activity_order.xml [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+        app:layout_constraintBaseline_toBaselineOf="@+id/address_label"
+        app:layout_constraintStart_toEndOf="@+id/address_label" />
+
+    <TextView
+        android:id="@+id/phone_label"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginLeft="24dp"
+        android:layout_marginStart="24dp"
+        android:layout_marginTop="24dp"
```
Comment (hungvv-1301) [inline app/src/main/res/layout/activity_order.xml]: .

File: app/src/main/res/layout/activity_order.xml [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+        android:text="@string/phone_label_string"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toBottomOf="@+id/address_text" />
+
+    <EditText
+        android:id="@+id/phone_text"
+        android:layout_width="134dp"
+        android:layout_height="wrap_content"
+        android:layout_marginLeft="8dp"
+        android:layout_marginStart="8dp"
```
Comment (hungvv-1301) [inline app/src/main/res/layout/activity_order.xml]: .

File: app/src/main/res/layout/activity_order.xml [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+        android:textSize="18sp"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toBottomOf="@+id/note_text" />
+
+    <RadioGroup
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginLeft="24dp"
+        android:layout_marginStart="24dp"
+        android:orientation="vertical"
```
Comment (hungvv-1301) [inline app/src/main/res/layout/activity_order.xml]: .

File: app/src/main/res/layout/activity_pick_date.xml [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+        android:id="@+id/button_time"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="@dimen/dp_15"
+        android:text="@string/time"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toBottomOf="@+id/button_date" />
+
+</androidx.constraintlayout.widget.ConstraintLayout>
```
Comment (hungvv-1301) [inline app/src/main/res/layout/activity_pick_date.xml]:
```suggestion
</androidx.constraintlayout.widget.ConstraintLayout>

```

File: app/src/main/res/menu/main_menu.xml [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+        app:showAsAction="always"
+        />
+    <item
+        android:id="@+id/action_favorites"
+        android:orderInCategory="30"
+        android:title="Favorites"
+        android:icon="@drawable/ic_favorites"
+        app:showAsAction="always"
+        />
+</menu>
```
Comment (hungvv-1301) [inline app/src/main/res/menu/main_menu.xml]:
```suggestion
</menu>

```

**Review by hungvv-1301 — COMMENTED** [review]
tất cả các file extract dp, sp , endline

File: app/src/main/java/com/sun/android/menus_picker/AlertActivity.kt [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+
+class AlertActivity : AppCompatActivity() {
+    val binding by lazy { ActivityAlertBinding.inflate(layoutInflater) }
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(binding.root)
+        binding.button.setOnClickListener{
+            var myAlertBuilder: AlertDialog.Builder = AlertDialog.Builder(this)
+            myAlertBuilder.setTitle("Alert")
+            myAlertBuilder.setMessage("Click OK to continue, or Cancel to stop: ")
```
Comment (hungvv-1301) [inline app/src/main/java/com/sun/android/menus_picker/AlertActivity.kt]: các biến k thay đổi thì dùng val , có thể dùng . để set các thuôc tính liên tiếp ( Builder partern)
```suggestion
            var myAlertBuilder: AlertDialog.Builder = AlertDialog.Builder(this)
           .setTitle("Alert")
           .setMessage("Click OK to continue, or Cancel to stop: ")
```

File: app/src/main/java/com/sun/android/menus_picker/MainActivityMenuPicker.kt [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    fun showFroyoOrder(view: View){
+        mOrderMessage = getString(R.string.froyo_order_message)
+        displayToast(mOrderMessage)
+    }
+
+
+
+
```
Comment (hungvv-1301) [inline app/src/main/java/com/sun/android/menus_picker/MainActivityMenuPicker.kt]:
```suggestion
```

File: app/src/main/java/com/sun/android/menus_picker/TimePickerFragment.kt [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+
+
+class TimePickerFragment : DialogFragment(), TimePickerDialog.OnTimeSetListener {
+    override fun onTimeSet(timePicker: TimePicker?, hourOfDay: Int, minute: Int) {
+        var activity = activity as PickDateActivity
+        activity.processTimePickerResult(hour = hourOfDay, minute = minute)
+    }
+
+    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
+        var c: Calendar = Calendar.getInstance()
```
Comment (hungvv-1301) [inline app/src/main/java/com/sun/android/menus_picker/TimePickerFragment.kt]:
```suggestion
        val calendar = Calendar.getInstance()
```

File: app/src/main/java/com/sun/android/menus_picker/PickDateActivity.kt [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+import androidx.fragment.app.DialogFragment
+import com.sun.android.databinding.ActivityPickDateBinding
+
+class PickDateActivity : AppCompatActivity() {
+    val binding by lazy { ActivityPickDateBinding.inflate(layoutInflater) }
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(binding.root)
+        binding.buttonDate.setOnClickListener{
+            var newFragmentDate: DialogFragment = DatePickerFragment()
```
Comment (hungvv-1301) [inline app/src/main/java/com/sun/android/menus_picker/PickDateActivity.kt]:
```suggestion
            val newFragmentDate = DatePickerFragment()
```

File: app/src/main/java/com/sun/android/menus_picker/PickDateActivity.kt [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+            newFragmentDate.show(supportFragmentManager,"datePicker")
+        }
+        binding.buttonTime.setOnClickListener{
+            var newFragmentTime: DialogFragment = TimePickerFragment()
```
Comment (hungvv-1301) [inline app/src/main/java/com/sun/android/menus_picker/PickDateActivity.kt]: .

File: app/src/main/java/com/sun/android/menus_picker/TimePickerFragment.kt [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+import android.widget.TimePicker
+import androidx.annotation.NonNull
+import androidx.fragment.app.DialogFragment
+import java.util.*
+
+
+class TimePickerFragment : DialogFragment(), TimePickerDialog.OnTimeSetListener {
+    override fun onTimeSet(timePicker: TimePicker?, hourOfDay: Int, minute: Int) {
+        var activity = activity as PickDateActivity
+        activity.processTimePickerResult(hour = hourOfDay, minute = minute)
```
Comment (hungvv-1301) [inline app/src/main/java/com/sun/android/menus_picker/TimePickerFragment.kt]:
```suggestion
        val activity = activity as? PickDateActivity
        activity?.processTimePickerResult(hour = hourOfDay, minute = minute)
```

File: app/src/main/java/com/sun/android/menus_picker/DatePickerFragment.kt [hunk 22]
```diff
@@ -0,0 +1,25 @@
...
+
+class DatePickerFragment : DialogFragment(), DatePickerDialog.OnDateSetListener {
+    override fun onDateSet(datePicker: DatePicker?, year: Int, month: Int, day: Int) {
+        var activity = activity as PickDateActivity
+        activity.processDatePickerResult(year = year, month = month, day = day)
+    }
+
+    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
+        var c: Calendar = Calendar.getInstance()
+        val year: Int = c.get(Calendar.YEAR)
```
Comment (ghost) [inline app/src/main/java/com/sun/android/menus_picker/DatePickerFragment.kt]:
```suggestion
        var calendar: Calendar = Calendar.getInstance()
        val year: Int = c.get(Calendar.YEAR)
```

File: app/src/main/java/com/sun/android/menus_picker/PickDateActivity.kt [hunk 23]
```diff
@@ -0,0 +1,33 @@
...
+        }
+        binding.buttonTime.setOnClickListener{
+            var newFragmentTime = TimePickerFragment()
+            newFragmentTime.show(supportFragmentManager,"timePicker")
+        }
+
+    }
+    fun processDatePickerResult(year: Int, month: Int, day: Int){
+        val dateMessage = "Date: $day/$month/$year"
+        Toast.makeText(this,  dateMessage, Toast.LENGTH_SHORT).show()
```
Comment (ghost) [inline app/src/main/java/com/sun/android/menus_picker/PickDateActivity.kt]: dùng string resource

File: app/src/main/java/com/sun/android/menus_picker/PickDateActivity.kt [hunk 24]
```diff
@@ -0,0 +1,33 @@
...
+    }
+
+    fun processTimePickerResult(hour: Int, minute: Int){
+        val timeMessage = " Time: $hour:$minute"
+        Toast.makeText(this,  timeMessage, Toast.LENGTH_SHORT).show()
```
Comment (ghost) [inline app/src/main/java/com/sun/android/menus_picker/PickDateActivity.kt]: .

File: app/src/main/res/layout/content_main_menu_picker.xml [hunk 25]
```diff
@@ -0,0 +1,97 @@
...
+
+    <TextView
+        android:id="@+id/textintro"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="8dp"
+        android:text="@string/intro_text"
+        android:textSize="24sp"
+        android:textStyle="bold"
+        app:layout_constraintLeft_toLeftOf="parent"
```
Comment (ghost) [inline app/src/main/res/layout/content_main_menu_picker.xml]: dùng file dimen

File: app/src/main/res/layout/content_main_menu_picker.xml [hunk 26]
```diff
@@ -0,0 +1,97 @@
...
+        app:layout_constraintRight_toRightOf="parent"
+        app:layout_constraintTop_toTopOf="parent" />
+
+    <ImageView
+        android:id="@+id/donut"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginLeft="24dp"
+        android:layout_marginStart="24dp"
+        android:layout_marginTop="24dp"
```
Comment (ghost) [inline app/src/main/res/layout/content_main_menu_picker.xml]: .

File: app/src/main/res/layout/content_main_menu_picker.xml [hunk 27]
```diff
@@ -0,0 +1,97 @@
...
+        app:layout_constraintTop_toBottomOf="@+id/textintro"
+        app:srcCompat="@drawable/donut_circle" />
+
+    <ImageView
+        android:id="@+id/ice_cream"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginLeft="24dp"
+        android:layout_marginStart="24dp"
+        android:layout_marginTop="24dp"
```
Comment (ghost) [inline app/src/main/res/layout/content_main_menu_picker.xml]: .

File: app/src/main/res/layout/content_main_menu_picker.xml [hunk 28]
```diff
@@ -0,0 +1,97 @@
...
+        app:srcCompat="@drawable/icecream_circle" />
+
+    <ImageView
+        android:id="@+id/froyo"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginLeft="24dp"
+        android:layout_marginStart="24dp"
+        android:layout_marginTop="24dp"
+        android:contentDescription="@string/froyo"
```
Comment (ghost) [inline app/src/main/res/layout/content_main_menu_picker.xml]: .

File: app/src/main/res/layout/content_main_menu_picker.xml [hunk 29]
```diff
@@ -0,0 +1,97 @@
...
+        app:layout_constraintTop_toBottomOf="@+id/ice_cream"
+        app:srcCompat="@drawable/froyo_circle" />
+
+    <TextView
+        android:id="@+id/donut_description"
+        android:layout_width="0dp"
+        android:layout_height="wrap_content"
+        android:layout_marginEnd="24dp"
+        android:layout_marginStart="24dp"
+        android:layout_marginTop="24dp"
```
Comment (ghost) [inline app/src/main/res/layout/content_main_menu_picker.xml]: .

File: app/src/main/res/layout/content_main_menu_picker.xml [hunk 30]
```diff
@@ -0,0 +1,97 @@
...
+        app:layout_constraintStart_toEndOf="@+id/donut"
+        app:layout_constraintTop_toTopOf="@+id/donut" />
+
+    <TextView
+        android:id="@+id/ice_cream_description"
+        android:layout_width="0dp"
+        android:layout_height="wrap_content"
+        android:layout_marginEnd="24dp"
+        android:layout_marginStart="24dp"
+        android:layout_marginTop="24dp"
```
Comment (ghost) [inline app/src/main/res/layout/content_main_menu_picker.xml]: .

File: app/src/main/res/layout/content_main_menu_picker.xml [hunk 31]
```diff
@@ -0,0 +1,97 @@
...
+        app:layout_constraintStart_toEndOf="@+id/ice_cream"
+        app:layout_constraintTop_toTopOf="@+id/ice_cream" />
+
+    <TextView
+        android:id="@+id/froyo_description"
+        android:layout_width="0dp"
+        android:layout_height="wrap_content"
+        android:layout_marginEnd="24dp"
+        android:layout_marginStart="24dp"
+        android:layout_marginTop="24dp"
```
Comment (ghost) [inline app/src/main/res/layout/content_main_menu_picker.xml]: .

**Pull request merged**